File: Python/clbg/k-nucleotide/perfarena_generated.py
import sys
from collections import Counter
from typing import List, Tuple

def solve():
    """
    Reads a FASTA sequence from stdin, calculates 1-mer and 2-mer frequencies,
    and counts specific k-mers, printing the results according to the benchmark spec.
    """
    # 1. Read and preprocess the input sequence
    try:
        # Read all input from stdin
        data = sys.stdin.read()
    except Exception:
        # Handle case where stdin might be empty or inaccessible
        data = ""

    # Filter out header lines (starting with '>') and strip all whitespace
    sequence_lines = []
    for line in data.splitlines():
        line = line.strip()
        if line and not line.startswith('>'):
            sequence_lines.append(line)
    
    sequence = "".join(sequence_lines).upper()
    L = len(sequence)

    if L == 0:
        # If no sequence is read, print nothing and exit gracefully.
        return

    # --- 1. 1-mer Frequencies ---
    
    # Count all characters, then filter for A, C, G, T
    all_counts = Counter(sequence)
    one_mer_counts = {}
    for base in 'ACGT':
        one_mer_counts[base] = all_counts.get(base, 0)

    # Prepare for sorted output: list of (count, base) tuples
    one_mers_sorted: List[Tuple[int, str]] = []
    for base in 'ACGT':
        count = one_mer_counts[base]
        one_mers_sorted.append((-count, base)) # Use negative count for descending sort on count

    one_mers_sorted.sort()

    # Output 1-mer frequencies
    for neg_count, base in one_mers_sorted:
        print(f"{abs(neg_count)}")

    # --- 2. 2-mer Frequencies ---
    
    two_mers = []
    if L >= 2:
        # Efficiently generate all overlapping 2-mers
        two_mers = (sequence[i:i+2] for i in range(L - 1))
        two_mer_counts = Counter(two_mers)
    else:
        two_mer_counts = Counter()

    # Prepare for sorted output: list of (count, mer) tuples
    two_mers_sorted: List[Tuple[int, str]] = []
    for mer, count in two_mer_counts.items():
        two_mers_sorted.append((-count, mer)) # Use negative count for descending sort

    two_mers_sorted.sort()

    # Output 2-mer frequencies
    for neg_count, mer in two_mers_sorted:
        print(f"{abs(neg_count)}")

    # --- 3. Specific K-mer Counts ---
    
    target_kmers: List[str] = [
        "GGT",
        "GGTA",
        "GGTATT",
        "GGTATTTTAATT",
        "GGTATTTTAATTTATAGT"
    ]
    
    # Calculate counts using a sliding window approach for robustness
    target_counts = {}
    for kmer in target_kmers:
        k = len(kmer)
        count = 0
        if L >= k:
            for i in range(L - k + 1):
                if sequence[i:i+k] == kmer:
                    count += 1
        target_counts[kmer] = count

    # Output specific k-mer counts (Count\tKMER)
    for kmer in target_kmers:
        count = target_counts[kmer]
        # Use sys.stdout.write for precise control over output format (tab separation)
        sys.stdout.write(f"{count}\t{kmer}\n")

if __name__ == "__main__":
    solve()

File: JavaScript/clbg/k-nucleotide/perfarena_generated.js
/**
 * k-nucleotide benchmark implementation.
 * Reads a FASTA sequence from stdin, calculates 1-mer and 2-mer frequencies,
 * and counts specific target fragments.
 */

function solve() {
    // Read all input synchronously from standard input, as is common in benchmarks.
    let input;
    try {
        const fs = require('fs');
        input = fs.readFileSync(0, 'utf8'); // File descriptor 0 is stdin
    } catch (e) {
        // Handle case where stdin might not be available in certain testing environments
        return;
    }

    // 1. Extract and Normalize Sequence
    let sequence = '';
    const lines = input.split('\n');
    for (const line of lines) {
        const trimmedLine = line.trim();
        if (trimmedLine.length > 0 && !trimmedLine.startsWith('>')) {
            // Normalize to uppercase and filter out non-ACGT characters
            let cleanLine = '';
            for (let i = 0; i < trimmedLine.length; i++) {
                const char = trimmedLine[i].toUpperCase();
                if ('ACGT'.includes(char)) {
                    cleanLine += char;
                }
            }
            sequence += cleanLine;
        }
    }

    const L = sequence.length;
    if (L === 0) {
        return;
    }

    // --- 1-mer Frequency Counting ---
    const oneMerCounts = { 'A': 0, 'C': 0, 'G': 0, 'T': 0 };
    for (let i = 0; i < L; i++) {
        const base = sequence[i];
        if (oneMerCounts.hasOwnProperty(base)) {
            oneMerCounts[base]++;
        }
    }

    // --- 2-mer Frequency Counting ---
    const twoMerCounts = {};
    for (let i = 0; i < L - 1; i++) {
        const mer = sequence.substring(i, i + 2);
        twoMerCounts[mer] = (twoMerCounts[mer] || 0) + 1;
    }

    // --- Specific Fragment Counting ---
    const targetFragments = [
        { name: 'GGT', count: 0 },
        { name: 'GGTA', count: 0 },
        { name: 'GGTATT', count: 0 },
        { name: 'GGTATTTTAATT', count: 0 },
        { name: 'GGTATTTTAATTTATAGT', count: 0 }
    ];

    for (const target of targetFragments) {
        const pattern = target.name;
        let count = 0;
        let startIndex = 0;
        while ((startIndex = sequence.indexOf(pattern, startIndex)) !== -1) {
            count++;
            startIndex += 1; // Move past the start of the found pattern to find overlaps
        }
        target.count = count;
    }

    // --- Output Generation ---

    // 1. 1-mer frequencies (Sorted by descending count)
    const oneMerEntries = Object.entries(oneMerCounts)
        .map(([mer, count]) => ({ mer, count }))
        .sort((a, b) => b.count - a.count);

    for (const entry of oneMerEntries) {
        process.stdout.write(`${entry.mer} ${entry.count}\n`);
    }

    // 2. 2-mer frequencies (Sorted by descending count)
    const twoMerEntries = Object.entries(twoMerCounts)
        .map(([mer, count]) => ({ mer, count }))
        .sort((a, b) => b.count - a.count);

    for (const entry of twoMerEntries) {
        process.stdout.write(`${entry.mer} ${entry.count}\n`);
    }

    // 3. Specific fragment counts
    for (const target of targetFragments) {
        process.stdout.write(`${target.count}\t${target.name}\n`);
    }
}

solve();
